keps/sig-api-machinery/storage-migration-auto-trigger.md

```diff
@@ -0,0 +1,148 @@
+---
+title: Automated Storage Version Migration with Storage Version Hash
+authors:
+  - "@xuchao"
+owning-sig: sig-api-machinery
+reviewers:
+  - "@deads2k"
+  - "@yliaog"
+approvers:
+  - "@deads2k"
+  - "@lavalamp"
+creation-date: 2019-01-23
+last-updated: 2019-01-23
+status: provisional
+---
+
+## Table of Contents
+
+* [Goal](#goal)
+* [API design](#api-design)
+* [Storage migration triggering controller](#storage-migration-triggering-controller)
+* [Implications to cluster operators](#implications-to-cluster-operators)
+* [Future work: HA clusters](#future-work-ha-clusters)
+
+## Goal
+
+As the discovery document now exposes the [storage version hash][], storage version
+migration should start automatically when the hash changes.
+
+[storage version hash]:https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/35-storage-version-hash.md
+
+## API design
+
+We introduce the `StorageState` API, as a CRD defined in the `migration.k8s.io`
+group.
+
+```golang
+// StorageState is the state of the storage, for a specific resource.
+type StorageState struct {
+  metav1.TypeMeta
+  // The name is "<resource>.<group>". The API validation will enforce it by
+  // comapring the name with the spec.resource.
+  metav1.ObjectMeta
+  Spec StorageStateSpec
+  Status StorageStateStatus
+}
+
+type StorageStateSpec {
+  // The resource this StoageState is about.
+  Resource GroupResource
+}
+
+type StorageStateStatus {
+  // The hash values of versions that the persisted objects might be encoded in.
+  // If empty, it means the stored versions are unknown.
```

Comment:
Who uses this field and why? When will it have more than one value? Is it inclusive or exclusive of the next field?

Reply:
Largely rewrote the comment on this field. PTAL.